modules/text-input/react/lib/TextInput.tsx

```diff
@@ -72,10 +72,36 @@ export default class TextInput extends React.Component<TextInputProps> {
 
   static defaultProps = {
     type: 'text',
+    inputRef: React.createRef(),
   };
+  componentDidMount() {
+    if (this.props.value) {
+      this.updateNativeInputValue();
+    }
+  }
+  componentDidUpdate(nextProps: TextInputProps) {
+    if (nextProps.value !== this.props.value) {
+      this.updateNativeInputValue();
+    }
+  }
+
+  updateNativeInputValue = () => {
+    // Changing value prop programmatically doesn't fire an Synthetic event or trigger native onChange.
+    // We also can not just update .value= here because React library overrides input value setter
+    // but we can call the function directly on the input as context.
+    // This will onChange events to fire no matter how value is updated.
+    if (this.props.inputRef && this.props.inputRef.current) {
+      const nativeInputValue = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(this.props.inputRef.current), 'value')
+      if (nativeInputValue && nativeInputValue.set) {
+        nativeInputValue.set.call(this.props.inputRef.current, this.props.value);
+      }
+      const event = new Event('input', { bubbles: true });
+      this.props.inputRef.current.dispatchEvent(event);
```

Comment:
I don't mean passing a `value` prop without an `onChange`, I mean forcing an `onChange` to be called when the parent component changes the `value` prop

Reply:
Gotcha. So far in my testing everything has been working correctly. 
The main use case for this is being able to clear the input field via a button or keyboard shortcut and let any top level listeners know something changed. Without this any listeners you are using to set the autocomplete menu items won't know the field was changed.

Reply:
One more use case is to change the value of the input to the text of the autocomplete item on click

Reply:
I've solve those problems in the past by invoking the `onChange` from those user actions explicitly. This guarantees event handlers are only called due to events and not data changes. Otherwise some use-cases require special state tracking to tell the difference between handlers called from user events vs data changes.

Sometimes we violate this for easier data fetching like fetching data on mount of a component. A component mount isn't necessarily tied to a user event.

Always forcing events on any data change is probably easier since you don't have the explicitly call `onChange` for the times when you change the value, but it prevents accidental firings on `onChange` when it was not intended. The bugs can be subtle. It might manifest itself more when side-effects happen inside an `onChange` like fetching data from a server. You may find `onChange` being fired with data that is never actually rendered.

Reply:
Ok I will move this code into the combobox and explicitly call it when programmatically changing the data